### PR TITLE
matrix variate normal distribution

### PR DESCRIPTION
**Pull request by ssydasheng**
Add matrix variate normal distribution

### CODE REVIEW

**Comment by botev** [comment]
Do you mind adding the `path_derivative` as in https://github.com/thu-ml/zhusuan/pull/68. Essentially when you calculate `log_prob` if it is true you do stop gradients on the parameters in the calculation of `log_prob`.

**Comment by thjashin** [comment]
@botev Seems that it has been done. Did I miss something?

**Comment by botev** [comment]
That's interesting, I thought I did not see `self.path_param`, but indeed it is there.

File: zhusuan/distributions/multivariate.py [hunk 1]
```diff
@@ -952,3 +953,213 @@ def _prob(self, given):
...
+            [self._n_in if isinstance(self._n_in, int) else None])
+        self._u_tril.get_shape().assert_is_compatible_with(expected_u_shape)
+        expected_v_shape = self._mean.get_shape()[:-2].concatenate(
+            [self._n_out if isinstance(self._n_out, int) else None] * 2)
+        self._v_tril.get_shape().assert_is_compatible_with(expected_v_shape)
+        # Dynamic
+        expected_u_shape = tf.concat(
+            [tf.shape(self._mean)[:-1], [self._n_in]], axis=0)
+        actual_u_shape = tf.shape(self._u_tril)
+        msg = ['MultivariateNormalCholesky.u_tril should have compatible '
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: MultivariateNormalCholesky -> MatrixVariateNormalCholesky

**Comment by ssydasheng** [comment]
I will fix them according to your opinions.

File: zhusuan/distributions/multivariate.py [hunk 2]
```diff
@@ -952,3 +953,213 @@ def _prob(self, given):
...
+            len_ = len(t.get_shape())
+            return tf.transpose(t, [i for i in range(len_-2)]+[len_-1, len_-2])
+
+        batch_mean = tile(mean)
+        batch_u_tril = tile(u_tril)
+        batch_v_tril = tile(v_tril)
+        noise = tf.random_normal(tf.shape(batch_mean), dtype=self.dtype)
+        samples = batch_mean + \
+            tf.matmul(tf.matmul(batch_u_tril, noise),
+                      transpose_last2_dims(batch_v_tril))
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: use `tf.matrix_transpose` instead

File: zhusuan/distributions/multivariate.py [hunk 3]
```diff
@@ -952,3 +953,213 @@ def _prob(self, given):
...
+
+        def tile(t):
+            new_shape = tf.concat([[n_samples], tf.ones_like(tf.shape(t))], 0)
+            return tf.tile(tf.expand_dims(t, 0), new_shape)
+
+        def transpose_last2_dims(t):
+            len_ = len(t.get_shape())
+            return tf.transpose(t, [i for i in range(len_-2)]+[len_-1, len_-2])
+
+        batch_mean = tile(mean)
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: don't need to tile mean since it will broadcast.

File: zhusuan/distributions/multivariate.py [hunk 4]
```diff
@@ -952,3 +953,213 @@ def _prob(self, given):
...
 GumbelSoftmax = Concrete
+
+
+class MatrixVariateNormalCholesky(Distribution):
+    """
+    The class of maxtrix variate normal distribution, where covariances
+    :math:`U` and :math:`V` are parameterized with the lower triangular
+    matrix in Cholesky decomposition,
+
+        .. math :: L \\text{s.t.} LL^T = \Sigma.
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: should be replaced by u and v

File: zhusuan/distributions/multivariate.py [hunk 5]
```diff
@@ -952,3 +953,213 @@ def _prob(self, given):
...
+        def transpose_last2_dims(t):
+            len_ = len(t.get_shape())
+            return tf.transpose(
+                t, [i for i in range(len_-2)]+[len_-1, len_-2])
+
+        y = given - mean
+        y_with_last_dim_changed = tf.expand_dims(tf.ones(tf.shape(y)[:-1]), -1)
+        Lu, _ = maybe_explicit_broadcast(
+            u_tril, y_with_last_dim_changed,
+            'MultivariateNormalCholesky.u_tril', 'expand_dims(given, -1)')
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: why is this broadcast?

**Comment by thjashin** [comment]
@ssydasheng How is it going?

File: zhusuan/distributions/multivariate.py (hunk 5 above)
Comment (ssydasheng) [reply zhusuan/distributions/multivariate.py]: y can have 1 additional sample dimension.

**Comment by ssydasheng** [comment]
1, path_derivative already included
2, n_in, n_out --> n_row, n_col
3, use tf.matrix_transpose()
4, fix error message
5, add formulas for U, V individually in the init description
6, fix typos

File: zhusuan/distributions/multivariate.py [hunk 6]
```diff
@@ -952,3 +953,205 @@ def _prob(self, given):
...
+    .. math ::
+        L_u \\text{s.t.} L_u L_u^T = U,\\; L_v \\text{s.t.} L_v L_v^T = V
+
+    See :class:`~zhusuan.distributions.base.Distribution` for details.
+
+    :param mean: An N-D `float` Tensor of shape [..., n_row, n_col]. Each slice
+        [i, j, ..., k, :, :] represents the mean of a single matrix variate
+        normal distribution.
+    :param u_tril: An N-D `float` Tensor of shape [..., n_row, n_row].
+        Each slice [i, ..., k, :, :] represents the lower triangular matrix in
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: i -> i, j

File: zhusuan/distributions/multivariate.py [hunk 7]
```diff
@@ -952,3 +953,205 @@ def _prob(self, given):
...
+        the Cholesky decomposition of the among-row covariance of a single
+        distribution.
+    :param v_tril: An N-D `float` Tensor of shape [..., n_col, n_col].
+        Each slice [i, ..., k, :, :] represents the lower triangular matrix in
```
Comment (thjashin) [inline zhusuan/distributions/multivariate.py]: i -> i, j

File: zhusuan/distributions/multivariate.py (hunk 5 above)
Comment (thjashin) [reply zhusuan/distributions/multivariate.py]: I see. Though note that actually `given` can have >=1 additional sample dimensions.

**Pull request merged**